### PR TITLE
Fixed #29687 -- Allowed the test client to serialize list/tuple as JSON.

### PR DESCRIPTION
**Pull request by danpalmer**
This addresses https://code.djangoproject.com/ticket/29687#ticket

I'm not sure what the standard practice is for documentation here. Particularly, this feels borderline between a bug and a feature – technically the documentation is explicit about only accepting dicts:

> If you provide `content_type` as `application/json`, a `data` dictionary is serialized using `json.dumps`.

...however when doing a Django 2.1 upgrade I found this to be unexpected, as I removed all of our explicit JSON serialisation from our test client code and expected it to work, but we use JSON arrays in some places and they didn't work.

Should I just change the documentation to read that arrays are supported too, or should I document this as changed behaviour and a new feature? If so, should I add the `versionchanged` tag? Which version should I specify?

### CODE REVIEW

**Comment by adamchainz** [comment]
I commented on ticket, if we go ahead this should be a 2.2 feature, and probably it's not just arrays that should be supported.

File: django/test/client.py [hunk 1]
```diff
@@ -320,8 +320,14 @@ def _encode_json(self, data, content_type):
         Return encoded JSON if data is a dict and content_type is
         application/json.
         """
-        should_encode = JSON_CONTENT_TYPE_RE.match(content_type) and isinstance(data, dict)
-        return json.dumps(data, cls=self.json_encoder) if should_encode else data
+        should_encode = JSON_CONTENT_TYPE_RE.match(content_type)
```
Comment (timgraham) [inline django/test/client.py]: Simpler could be:
```
if JSON_CONTENT_TYPE_RE.match(content_type):
    try:
        return json.dumps(data, cls=self.json_encoder)
    except TypeError:
        pass
return data
```

File: docs/topics/testing/tools.txt [hunk 2]
```diff
@@ -226,6 +226,11 @@ Use the ``django.test.Client`` class to make requests.
             you can call :func:`json.dumps` on ``data`` before passing it to
             ``post()`` to achieve the same thing.
 
+        .. versionchanged:: 2.2
+
+            The JSON serialization was extended to support all types serializable
```
Comment (timgraham) [inline docs/topics/testing/tools.txt]: This note should be "self-contained" (i.e. the main paragraph above should reflect the new behavior and this note should say something to the effect of "In older versions, only data dictionaries are serialized."

File: tests/test_client/tests.py [hunk 3]
```diff
@@ -92,17 +93,46 @@ def test_post(self):
...
+            'True': (True, True),
+            'False': (False, False),
+            'None': (None, None),
+        }
+
+        def _test(method, example, expected):
+            client_method = getattr(self.client, method)
+            method_name = method.upper()
+
+            # Test lists
```
Comment (timgraham) [inline tests/test_client/tests.py]: obsolete

File: tests/test_client/tests.py [hunk 4]
```diff
@@ -92,17 +93,46 @@ def test_post(self):
...
+            response = client_method(
```
Comment (timgraham) [inline tests/test_client/tests.py]: This can be a single line.

File: tests/test_client/tests.py [hunk 5]
```diff
@@ -92,17 +93,46 @@ def test_post(self):
...
+                '/json_view/',
+                example,
```
Comment (timgraham) [inline tests/test_client/tests.py]: example -> data

File: tests/test_client_regress/tests.py [hunk 6]
```diff
@@ -1189,11 +1189,11 @@ def test_patch(self):
     def test_empty_string_data(self):
         "Request a view with empty string data via request method GET/POST/HEAD"
         # Regression test for #21740
-        response = self.client.get('/body/', data='', content_type='application/json')
+        response = self.client.get('/body/', data='', content_type='text/plain')
```
Comment (timgraham) [inline tests/test_client_regress/tests.py]: What's the rationale for this change? See https://code.djangoproject.com/ticket/21740#comment:6 for the original regression.

File: tests/test_client/tests.py [hunk 7]
```diff
@@ -92,17 +93,46 @@ def test_post(self):
...
+            IntValue = 1
+
+        python_types_expected_values = {
+            'dict': ({'value': 37}, {'value': 37}),
+            'list': ([37], [37]),
+            'tuple': ((37,), [37]),
+            'str': ("Value", "Value"),
+            'int': (37, 37),
+            'float': (37.0, 37.0),
+            'int_enum': (IntEnum.IntValue, 1),
```
Comment (timgraham) [inline tests/test_client/tests.py]: I think a few basic types are enough to test. This isn't testing anything particular in Django as far as I can tell.

File: tests/test_client/tests.py (hunk 5 above)
Comment (danpalmer) [reply tests/test_client/tests.py]: 919903898b

File: tests/test_client/tests.py (hunk 3 above)
Comment (danpalmer) [reply tests/test_client/tests.py]: 919903898b

File: tests/test_client/tests.py (hunk 4 above)
Comment (danpalmer) [reply tests/test_client/tests.py]: 78ad571919

File: django/test/client.py (hunk 1 above)
Comment (danpalmer) [reply django/test/client.py]: 919903898b

**Comment by timgraham** [comment]
You can push a single squashed commit. The patch is simple enough that it's not aiding review to have multiple commits and GitHub sends a notification for every commit and comment.

File: docs/topics/testing/tools.txt (hunk 2 above)
Comment (danpalmer) [reply docs/topics/testing/tools.txt]: 384455f1e2

File: tests/test_client/tests.py (hunk 7 above)
Comment (danpalmer) [reply tests/test_client/tests.py]: Fair enough. Given that they are there I think we might as well keep them. It makes it less likely that someone would try to guess all the types for an `isinstance` check.

File: tests/test_client_regress/tests.py (hunk 6 above)
Comment (danpalmer) [reply tests/test_client_regress/tests.py]: So the rationale here is that...

 - That given example does not appear to rely on being JSON specifically – it's just an example.
 - The expected serialisation for `''` with a JSON content type is `b'""'`, not `b''`.
 - We can't drop the `content_type` parameter here because it defaults to multipart data, which won't serialise a string.

This is a behaviour change, but one that makes the behaviour more uniform given the expectation that JSON content is serialised automatically.

**Comment by danpalmer** [comment]
Just squashed everything, although I've kept the commit message detailing why I changed the regression test.

File: tests/test_client_regress/tests.py (hunk 6 above)
Comment (timgraham) [reply tests/test_client_regress/tests.py]: My impression is that the example from the comment was actual code from the commenter's project. I'm not sure if a string should be encoded again as this would be backwards incompatible (I think) if someone is already encoding their data pre-Django 2.1 (i.e. using the advice in the documentation "Use json.dumps() on data before passing it to post()" (I didn't test it, so let me know if that's a wrong assumption).

Comment (danpalmer) [reply tests/test_client_regress/tests.py]: Hmmm. Good point, that isn't backwards incompatible in that case. Do you think it's best to special case strings in that case?

To do that, I would:

 - Add a special case
 - Remove the string test
 - Update the docs
 - Revert the regression test change
 - Add another regression test with an explicit example of a json dumped string.

Comment (danpalmer) [reply tests/test_client_regress/tests.py]: Alternatively, we could leave this one entirely and create a new optional `json` parameter to these functions, that sets the content type by default, and encodes everything passed to it as JSON.

Thoughts?

Comment (danpalmer) [reply tests/test_client_regress/tests.py]: I've gone ahead and implemented the first option here, as the second would probably be a bigger discussion. Let me know if you want me to do the latter though – I'd do it in a separate PR, but I'd be happy to do it.

File: django/test/client.py [hunk 8]
```diff
@@ -320,8 +320,12 @@ def _encode_json(self, data, content_type):
         Return encoded JSON if data is a dict and content_type is
         application/json.
         """
-        should_encode = JSON_CONTENT_TYPE_RE.match(content_type) and isinstance(data, dict)
-        return json.dumps(data, cls=self.json_encoder) if should_encode else data
+        if JSON_CONTENT_TYPE_RE.match(content_type) and not isinstance(data, str):
+            try:
+                return json.dumps(data, cls=self.json_encoder)
+            except TypeError:
+                pass
```
Comment (orf) [inline django/test/client.py]: I feel like this should raise an exception and not be swallowed silently:

```
In [1]: class Abc:
   ...:     pass

In [2]: import json

In [3]: json.dumps([Abc()])
...
TypeError: Object of type Abc is not JSON serializable
```

If `[Abc()]` is passed here it would result in `[Abc()]` being returned, which might error somewhere down the line in a weird fashion, or worse end up with `str([Abc()])` being sent.

Comment (danpalmer) [reply django/test/client.py]: I thought about this, and happy to go with that, but I thought that falling back to the rest of the data handling logic applying as normal could be also considered a good thing.

For example, currently using `data=json.dumps({}).encode('utf-8')` would work, but if this threw the `TypeError` it would fail, certainly breaking backwards compatibility with our tests, and possibly with other use-cases.

I'm planning to leave as-is for this reason, but do let me know if you feel strongly enough! If so I'd appreciate advice on how this should be documented.

Comment (adamchainz) [reply django/test/client.py]: > currently using `data=json.dumps({}).encode('utf-8')` would work

Yes but `bytes` as body is supported as the default for other content types, and `bytes` aren't JSON serializable either. I think if you ignored both `str` and `bytes` on 323 then we could probably raise the exceptions fine?

> certainly breaking backwards compatibility with our tests

"our" = Django's or Thread's? Have you got any other examples?


Alternatively you could be changing the logic on line 323 to increase the whitelist of allowed types - i.e. add `list` and perhaps `int` / `float` / `bool` - rather than changing it to a blacklist of types *not* to `json.dumps`. Granted there may be edge cases for other json encoders, like `Decimal`, but it's safer for everyone's tests, and more obvious than swallowing an exception.

Comment (timgraham) [reply django/test/client.py]: I feel similarly. Maybe it's better just to add support for list and tuple as originally proposed. It's unclear to me if other types would actually be used.

Comment (adamchainz) [reply django/test/client.py]: Only `list` and `tuple` only would be fine by me, you're right it's unclear the other default json types would be used.

Comment (danpalmer) [reply django/test/client.py]: Great. I don't feel strongly. I'll roll this back to just list/tuple/dict and update.

**Comment by danpalmer** [comment]
Right, I've updated the PR to:

 - Only support JSON objects and lists – the most expected use-cases, means we don't need to deal with edge cases around string/bytes support.
 - Not to swallow exceptions – this was mostly supporting the serialisation of arbitrary content types based on the provided encoder, but we aren't doing this anymore so we shouldn't silence potentially useful exceptions.
 - Updated docs for the above.

I think this addresses all of the code review given. Happy to address any other concerns!
